### PR TITLE
Add logging callback to DAC EnumMemoryRegion API

### PR DESCRIPTION
**Pull request by mikem8361**
Add the ICLRDataEnumMemoryRegionsLoggingCallback interface so createdump can receive logging from the EnumMemoryRegion API.

Add logging to MethodTable EnumMemoryRegion for invalid EEClass/Canonical MT.

Add checked build logging after waitpid for createdump to help diagnose issue https://github.com/dotnet/runtime/issues/72755.

### CODE REVIEW

**Comment by mikem8361** [comment]
@noahfalk, @hoyosjs I would like to get this PR in before the end of the week.

File: src/coreclr/vm/threads.cpp [hunk 1]
```diff
@@ -8459,6 +8459,11 @@ Thread::EnumMemoryRegionsWorker(CLRDataEnumMemoryFlags flags)
     }
     else
     {
+        // Skip any thread that doesn't have a OS thread id because DacGetThreadContext is going to throw an exception.
+        if (GetOSThreadId() == 0)
+        {
+            return;
```
Comment (noahfalk) [inline src/coreclr/vm/threads.cpp:8465]: Now that we have a logging mechanism should we log a message about this?

File: src/coreclr/debug/daccess/dacimpl.h [hunk 2]
```diff
@@ -1337,6 +1337,19 @@ class ClrDataAccess
     bool ReportMem(TADDR addr, TSIZE_T size, bool fExpectSuccess = true);
     bool DacUpdateMemoryRegion(TADDR addr, TSIZE_T bufferSize, BYTE* buffer);
 
+    inline bool IsLogMessage()
```
Comment (noahfalk) [inline src/coreclr/debug/daccess/dacimpl.h:1340]:
```suggestion
    inline bool IsLogMessageEnabled()
```

File: src/coreclr/debug/createdump/crashinfo.h [hunk 3]
```diff
@@ -35,7 +35,7 @@ extern const std::string GetDirectory(const std::string& fileName);
 extern std::string FormatString(const char* format, ...);
 extern std::string FormatGuid(const GUID* guid);
 
-class CrashInfo : public ICLRDataEnumMemoryRegionsCallback,
+class CrashInfo : public ICLRDataEnumMemoryRegionsCallback, public ICLRDataEnumMemoryRegionsLoggingCallback,
```
Comment (noahfalk) [inline src/coreclr/debug/createdump/crashinfo.h:38]:
```suggestion
class CrashInfo : public ICLRDataEnumMemoryRegionsCallback, public ICLRDataLoggingCallback,
```

The logging could also occur in scenarios that are not doing EnumMemoryRegions if we wanted to register the callback right? Even if we only register the logging callback during EnumMemoryRegions right now I'd suggest we pick a name that will still make sense if we make broader usage of this logging in the future.

**Review by noahfalk — APPROVED** [review]
LGTM aside from some small naming suggestions

File: src/coreclr/vm/threads.cpp (hunk 1 above)
Comment (mikem8361) [reply src/coreclr/vm/threads.cpp:8465]: This seems like a kind of normal condition to have no native thread id.  This check was to keep the GetThreadContext from throwing and logging a message. I wanted to keep the "exception thrown" logging only for actually problems.

So I don't think it needs logging.

File: src/coreclr/pal/prebuilt/idl/clrdata_i.cpp [hunk 4]
```diff
@@ -1,16 +1,13 @@
-// Licensed to the .NET Foundation under one or more agreements.
-// The .NET Foundation licenses this file to you under the MIT license.
+
```
Comment (hoyosjs) [inline src/coreclr/pal/prebuilt/idl/clrdata_i.cpp:1]: Please add the license again.

```suggestion
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
```

File: src/coreclr/pal/prebuilt/inc/clrdata.h [hunk 5]
```diff
@@ -1,14 +1,11 @@
-// Licensed to the .NET Foundation under one or more agreements.
-// The .NET Foundation licenses this file to you under the MIT license.
+
```
Comment (hoyosjs) [inline src/coreclr/pal/prebuilt/inc/clrdata.h:1]:
```suggestion
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
```

**Review by hoyosjs — APPROVED** [review]

**Pull request merged**